server/bleep/src/agent/symbol.rs

```diff
@@ -0,0 +1,370 @@
+            .hoverable_ranges()
+            .ok_or_else(|| anyhow::anyhow!("no hoverable ranges"))?;
+
+        let all_docs = {
+            let associated_langs = match document.lang.as_deref().map(TSLanguage::from_id) {
+                Some(Language::Supported(config)) => config.language_ids,
+                _ => &[],
+            };
+            self.app
+                .indexes
+                .file
+                .by_repo(&self.repo_ref, associated_langs.iter(), None)
+                .await
+        };
+
+        // get references and definitions for each symbol
+        let related_symbols = futures::future::join_all(
+            hoverable_ranges
+                .iter()
+                .filter(|range| {
+                    (range.start.byte >= chunk.start_byte) && (range.start.byte < chunk.end_byte)
+                })
+                .map(|range| {
+                    get_token_info(
+                        TokenInfoRequest {
+                            relative_path: chunk.path.clone(),
+                            repo_ref: String::new(), // FIXME
+                            branch: None,
+                            start: range.start.byte,
+                            end: range.end.byte,
+                        },
+                        &self.repo_ref,
+                        self.app.indexes.clone(),
+                        &document,
+                        &all_docs,
+                    )
+                }),
+        )
+        .await;
+
+        // filter references and definitions
+        let mut symbols = related_symbols
+            .into_iter()
+            .filter_map(Result::ok)
+            .zip(hoverable_ranges.into_iter().filter(|range| {
+                (range.start.byte >= chunk.start_byte) && (range.start.byte < chunk.end_byte)
+            }))
+            .map(|(token_info, range)| {
+                let filtered_token_info = token_info
+                    .into_iter()
+                    .filter(|file_symbols| file_symbols.file != chunk.path)
+                    .collect::<Vec<_>>();
+
+                Symbol {
+                    name: chunk.snippet.clone()[(range.start.byte - chunk.start_byte)
+                        ..(range.end.byte - chunk.start_byte)]
+                        .to_string(),
+                    related_symbols: filtered_token_info,
+                }
+            })
+            .filter(|metadata| {
+                (metadata.related_symbols.len() < MAX_REF_DEFS)
+                    && (!metadata.related_symbols.is_empty())
+            })
+            .collect::<Vec<_>>();
+
+        symbols.sort_by(|a, b| a.name.cmp(&b.name));
+        symbols.dedup_by(|a, b| a.name == b.name);
+
+        debug!("Attatched {} symbols", symbols.len());
+
+        Ok(ChunkWithSymbols {
+            chunk: chunk.clone(),
+            symbols,
+        })
+    }
+
+    pub async fn expand_symbol_into_chunks(&self, ref_def_metadata: Symbol) -> Vec<CodeChunk> {
```

Comment:
@nerdypepper Is there some logic that we can reuse for this?

Reply:
I changed token-info so we can define `context_before` and `context_after`